Enforce reserve when creating trust line or MPToken in `VaultWithdraw`

Bronek opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

Similarly to other transaction typed which can create a trust line or MPToken for the transaction submitter (e.g. `CashCheck` #5285 , `EscrowFinish` #5185 ), `VaultWithdraw` should enforce reserve before creating a new object.

Additionally, enforce `lsfRequireDestTag` account flag for the transaction submitter.

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

This fixes #5837 and improves consistency with other transaction types.

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [x] Bug fix (non-breaking change which fixes an issue)

dangell7 commented on src/test/app/Vault_test.cpp:
```diff
@@ -2659,6 +2696,64 @@ class Vault_test : public beast::unit_test::suite
+                env(tx2);
+                env.close();
+            },
+            CaseArgs{.initialXRP = 350});
```
use acctReserve + (incReserve - 1) or something similar

https://github.com/XRPLF/rippled/blob/5d79bfc5317093b00b787dfd5e50ce3bee1de0ca/src/test/app/EscrowToken_test.cpp#L701

Bronek replied:
Thanks, done

dangell7 commented on src/test/app/Vault_test.cpp:
```diff
@@ -1888,6 +1888,37 @@ class Vault_test : public beast::unit_test::suite
                     env(tx, ter(tecNO_AUTH));
```
It's hard to tell with your test case/suite structure but do you have tests where the issuer is the depositor? Also you test frozen but I don't see deep freeze or auth or no ripple?

Bronek replied:
In this branch, there is withdrawal to 3rd party which happens to be issuer  https://github.com/XRPLF/rippled/blob/fd178c550b434534209d84e700ffa0c7ebdfca74/src/test/app/Vault_test.cpp#L376

However this branch does not allow the issuer to be also depositor. This is fixed in #5518 and the test to match is https://github.com/XRPLF/rippled/blob/42be446f6979299a2c4eee58514c823b578af4ac/src/test/app/Vault_test.cpp#L399

ximinez commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +313,28 @@ VaultWithdraw::doApply()
+    if (!vaultAsset.native() && vaultAsset.holds<Issue>() &&
+        dstAcct != vaultAsset.getIssuer())
+    {
+        auto const sleAcct = view().read(keylet::account(dstAcct));
+        if (!sleAcct)
+            return tecINTERNAL;  // LCOV_EXCL_LINE
+
+        if (auto const sleLine =
+                view().read(keylet::line(dstAcct, vaultAsset.get<Issue>()));
+            sleLine == nullptr)
+        {
+            // This should be enforced by `requireAuth` called in `preclaim`
+            if (dstAcct != account_)
+                return tecINTERNAL;  // LCOV_EXCL_LINE
+
+            // Can the account cover the trust line's reserve ?
+            std::uint32_t const ownerCount = sleAcct->at(sfOwnerCount);
+            if (mPriorBalance < view().fees().accountReserve(ownerCount + 1))
+                return tecNO_LINE_INSUF_RESERVE;
+        }
```
I'm going to suggest a little reorganization here.

```suggestion
        if (auto const sleLine =
                view().read(keylet::line(dstAcct, vaultAsset.get<Issue>()));
            sleLine == nullptr)
        {
            // This should be enforced by `requireAuth` called in `preclaim`
            if (dstAcct != account_)
                return tecINTERNAL;  // LCOV_EXCL_LINE

            auto const sleAcct = view().read(keylet::account(account_));
            if (!sleAcct)
                return tecINTERNAL;  // LCOV_EXCL_LINE

            // Can account_ cover the trust line's reserve ?
            std::uint32_t const ownerCount = sleAcct->at(sfOwnerCount);
            if (mPriorBalance < view().fees().accountReserve(ownerCount + 1))
                return tecNO_LINE_INSUF_RESERVE;
        }
```

Basically, don't load the `AccountRoot` `SLE` until after you've verified that the line doesn't exist, and then when you do, specify it as `account_`, so it's clearer which account you're dealing with.

Even though computing the `Keylet` for the line is slightly more expensive than computing the `Keylet` for the account, if the line already exists, or if `dstAcct != account_`, you'll skip computing the account key entirely, and loading the object entirely.

Bronek replied:
Thanks, this is good.

ximinez commented on src/test/app/Vault_test.cpp:
```diff
@@ -1888,6 +1888,37 @@ class Vault_test : public beast::unit_test::suite
+                         .id = keylet.key,
+                         .amount = asset(100)});
+                    tx[sfDestination] = owner.human();
+                    env(tx, ter(tecNO_AUTH));
```
Wouldn't hurt to `env.close()` here, too.

Bronek commented:
While writing this I've realised that `VaultWithdraw` should also create `MPToken` when the submitter is also the destination account, and the `MPToken` for an MPT asset does not exist but otherwise would be authorized. I will add this change here.

This is similar to what the `TokenEscrow` amendment has done for MPT in `EscrowFinish` transaction.

Bronek commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -144,6 +141,16 @@ VaultWithdraw::preclaim(PreclaimContext const& ctx)
+
+        if (sleAcct->isFlag(lsfRequireDestTag) &&
+            !ctx.tx.isFieldPresent(sfDestinationTag))
+            return tecDST_TAG_NEEDED;  // Cannot send without a tag
```
this change is for consistency with other transaction types

ximinez replied:
> this change is for consistency with other transaction types

This code is basically duplicated from the `account != dstAccount` block, other than the result for a missing account.

Would it make more sense to take this code out of the `if` statement, and check which account before deciding which `TER` to return?

Something like
```
    auto const sleDst = ctx.view.read(keylet::account(dstAcct));
    if (sleDst == nullptr)
        return account == dstAcct : tecINTERNAL : tecNO_DST;

    if (sleDst->isFlag(lsfRequireDestTag) &&
        !ctx.tx.isFieldPresent(sfDestinationTag))
        return tecDST_TAG_NEEDED;  // Cannot send without a tag

    // If sending to Account (i.e. not a transfer), we will also create
    // (only if authorized) a trust line or MPToken as needed, in doApply().
    AuthType authType = AuthType::WeakAuth;
    if (account != dstAcct)
    {
        if (sleDst->isFlag(lsfDepositAuth))
        {
            if (!ctx.view.exists(keylet::depositPreauth(dstAcct, account)))
                return tecNO_PERMISSION;
        }
        // The destination account must have consented to receive the asset by
        // creating a RippleState or MPToken
        authType = AuthType::StrongAuth;
    }

Bronek commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,64 @@ VaultWithdraw::doApply()
+                                    MPTokenAuthorize::createMPToken(
+                                        view(), issue.getMptID(), account_, 0);
+                                !isTesSuccess(ter))
+                                return ter;  // LCOV_EXCL_LINE
```
the only possible error here is `tecDIR_FULL` which we normally do not test for.

ximinez commented on src/test/app/Vault_test.cpp:
```diff
@@ -67,6 +67,7 @@ class Vault_test : public beast::unit_test::suite
                                       Account const& owner,
                                       Account const& depositor,
                                       Account const& charlie,
+                                      Account const& dave,
```
If this list of `Account` params ever gets too unwieldy, you can simply "recreate" them inside the function. The values are deterministic and cached, so you don't have to worry about wasting resources either.

ximinez replied:
> If this list of `Account` params ever gets too unwieldy, you can simply "recreate" them inside the function. The values are deterministic and cached, so you don't have to worry about wasting resources either.

I see you have already rewritten them to have the accounts be in an outer scope, which is even simpler.

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,66 @@ VaultWithdraw::doApply()
         return account_;
     }();
 
+    if (!vaultAsset.native() && dstAcct != vaultAsset.getIssuer())
```
Shouldn't also check if `dstAcct == account_`? A trusdtline or MPToken only should be created if that condition is true.

Bronek replied:
Good idea, I will do it.

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,66 @@ VaultWithdraw::doApply()
+
+                        // If the above succeeded then the required trust line
+                        // will be created indirectly by `accountSend` below.
+                        return tesSUCCESS;
```
Don't need this. Can just let it fall through.

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,66 @@ VaultWithdraw::doApply()
+
+                            // Above will perform reserve check, create MPToken,
+                            // attach it to acccount_, adjust owner count etc.
+                            return tesSUCCESS;
```
Don't need this. Can just let it fall through.

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,66 @@ VaultWithdraw::doApply()
     }();
 
+    if (!vaultAsset.native() && dstAcct != vaultAsset.getIssuer())
+    {
```
There is a lambda on line 317 to get `dstAcct`. It can be simplified perhaps:

    auto const dstAcct = ctx_.tx[~sfDestination].value_or(account_);

Bronek replied:
Good idea.

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -313,6 +320,66 @@ VaultWithdraw::doApply()
+                            sleLine == nullptr)
+                        {
+                            // Enforced by `requireAuth` called in `preclaim`
+                            if (dstAcct != account_)
```
`dstAcct` may be different from `account_`. I think should only execute `std::visit` if `dstAcct == account_`.

Bronek replied:
Agree, however this condition is not faulty - it just relies on check performed in `preclaim`. But I like your idea better.

gregtatcam commented on src/test/app/Vault_test.cpp:
```diff
@@ -67,6 +67,7 @@ class Vault_test : public beast::unit_test::suite
                                       Account const& owner,
                                       Account const& depositor,
                                       Account const& charlie,
+                                      Account const& dave,
```
I think can simplify `testSequence` and `testCases` by defining the accounts once at the start of `teseSequences`:

    void
    testSequences()
    {
        using namespace test::jtx;
         Account const issuer{"issuer"};
         Account const owner{"owner"};
         Account const depositor{"depositor"};
         Account const charlie{"charlie"};  // authorized 3rd party
         Account const dave{"dave"};

        auto const testSequence = [this](
                                      std::string const& prefix,
                                      Env& env,
                                      Vault& vault,
                                      PrettyAsset const& asset)

gregtatcam replied:
This could be true for some other code when the passed accounts are always the same. For instance, `issuer` is always `Account{"issuer"}'.

Bronek replied:
Good idea

Bronek commented:
@gregtatcam your feedback is addressed in 7a55863d0c5cc7c8af53506f47aed1ebc5251d66 - thank you !

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -307,11 +314,57 @@ VaultWithdraw::doApply()
+                    if constexpr (std::is_same_v<TIss, Issue>)
+                    {
+                        std::shared_ptr<SLE> dummy = {};
+                        if (auto const sleLine =
```
This is more efficient:

     if (!view().exists(keylet::line(dstAcct, issue)))

Bronek replied:
also more explicit - thanks !

gregtatcam commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -307,11 +314,57 @@ VaultWithdraw::doApply()
+                    }
+                    else
+                    {
+                        if (auto const sleMPT = view().read(
```
This is more efficient:

     if (!view().exists(keylet::mptoken(issue.getMptID(), account_)))

gregtatcam reviewed (approved):
👍 LGTM

dangell7 commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -307,11 +292,16 @@ VaultWithdraw::doApply()
+    {
+        if (auto const ter = addEmptyHolding(
+                view(), account_, mPriorBalance, vaultAsset, j_);
+            ter && ter != tecDUPLICATE)
```
This could/should return tesSUCCESS if the line already exists imo. 

Also you call isGlobalFrozen here but also call it in preclaim too right? `canAddHolding` does the same checks as this but I don't see `canAddHolding` called? Does that mean that this function can return internal?

ximinez replied:
> This could/should return tesSUCCESS if the line already exists imo.

There are callers where it would be a weird problem if the holding already exists. For example, the transactions that create new accounts.

Bronek replied:
Normally I would suggest an additional flag as a function input parameter to choose the semantics of duplicates handling. In this case, if we want to go this way, #5270 would be a better fit to make a change (I would not want this small PR to be an obstacle for the larger PR to trip on).

shawnxie999 commented on src/xrpld/app/tx/detail/VaultWithdraw.cpp:
```diff
@@ -307,11 +292,16 @@ VaultWithdraw::doApply()
+    {
+        if (auto const ter = addEmptyHolding(
+                view(), account_, mPriorBalance, vaultAsset, j_);
+            ter && ter != tecDUPLICATE)
```
nit:  I find it a bit more clear by explicitly specifying that we are looking for non-tesSuccess code
```suggestion
            !isTesSuccess(ter) && ter != tecDUPLICATE)
```

shawnxie999 reviewed (approved):
LGTM 👍

bthomee committed:
fix: Enforce reserve when creating trust line or MPToken in VaultWithdraw

Similarly to other transaction typed that can create a trust line or MPToken for the transaction submitter (e.g. CashCheck #5285, EscrowFinish #5185 ), VaultWithdraw should enforce reserve before creating a new object. Additionally, the lsfRequireDestTag account flag should be enforced for the transaction submitter.

ximinez reviewed (commented):
Post merge approval 👍
